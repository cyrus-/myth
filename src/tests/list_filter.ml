let s = "
type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

type bool =
  | True
  | False

let rec is_even (n:nat) : bool =
  match n with
  | O -> True
  | S (n1) ->
    match n1 with
    | O -> False
    | S (n2) -> is_even n2
;;

let rec is_nonzero (n:nat) : bool =
  match n with
  | O -> False
  | S (n1) -> True
;;

let list_filter : (nat -> bool) -> list -> list |>
{
  is_even => ( [] => []
             | [0] => [0]
             | [1] => []
             | [2] => [2]
             | [0;0] => [0;0]
             | [0;1] => [0] )
| is_nonzero => ( [] => []
                | [0] => [] )
} = ?"
