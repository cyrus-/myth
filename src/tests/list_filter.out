(* Generated Thu Nov 13 00:35:26 2014 *)
let list_filter : (nat -> bool) -> list -> list =
  fun (f2:nat -> bool) ->
    let rec f3 (l1:list) : list =
      match l1 with
        | Nil -> Nil
        | Cons (n1, l2) -> (match f2 n1 with
                              | True -> Cons (n1, f3 l2)
                              | False -> Nil)
    in
      f3
;;
